File: AN_2/SEM_2/MPP/c#/lab5/Server/mpp.server/ServicesImpl.cs
using System.Collections.Concurrent;

namespace Server.mpp.server;
using Model.mpp.model;
using Persistence.mpp.persistence;
using Services.mpp.services;

public class ServicesImpl : IServices
{
    private readonly ICursaRepository _cursaRepository;
    private readonly IEchipaRepository _echipaRepository;
    private readonly IParticipantRepository _participantRepository;
    private readonly IParticipantCursaRepository _participantCursaRepository;
    private readonly IUtilizatorRepository _utilizatorRepository;
    private readonly ConcurrentDictionary<string, List<IObserver>> loggedClients = new();

    public ServicesImpl(
        IUtilizatorRepository utilizatorRepository,
        IEchipaRepository echipaRepository,
        IParticipantRepository participantRepository,
        ICursaRepository cursaRepository,
        IParticipantCursaRepository participantCursaRepository)
    {
        _utilizatorRepository = utilizatorRepository;
        _echipaRepository = echipaRepository;
        _participantRepository = participantRepository;
        _cursaRepository = cursaRepository;
        _participantCursaRepository = participantCursaRepository;
    }

    public void Login(Utilizator utilizator, IObserver client)
    {
        if (utilizator == null)
            throw new Exception("Utilizator invalid.");

        lock (loggedClients)
        {
            if (!loggedClients.ContainsKey(utilizator.Nume))
            {
                loggedClients[utilizator.Nume] = new List<IObserver>();
            }

            var observers = loggedClients[utilizator.Nume];

            if (!observers.Contains(client))
            {
                observers.Add(client);
            }
            else
            {
                throw new Exception("Observer deja logat.");
            }
        }
    }
    
    public void Logout(Utilizator utilizator, IObserver client)
    {
        if (utilizator == null)
            return;

        lock (loggedClients)
        {
            if (loggedClients.TryGetValue(utilizator.Nume, out var observers))
            {
                observers.Remove(client);

                if (observers.Count == 0)
                {
                    loggedClients.TryRemove(utilizator.Nume, out _);
                }
            }
        }

        Console.WriteLine($"{utilizator.Nume} s-a delogat (observer: {client.GetType().Name})");
    }

    public Participant FindByNume(string nume)
    {
        return _participantRepository.FindByNume(nume);
    }
    
    public void InscriereParticipant(Participant participant)
    {
        _participantRepository.Adauga(participant);
    }

    public List<Participant> FindAllParticipanti()
    {
        return _participantRepository.FindAll();
    }

    public List<Participant> CautaParticipantiDupaEchipa(string numeEchipa)
    {
        var echipa = _echipaRepository.FindByNume(numeEchipa);
        if (echipa == null)
            throw new Exception("Echipa nu exista");

        return _participantRepository.FindByEchipa(echipa);
    }

    public Echipa FindEchipaByNume(string nume)
    {
        return _echipaRepository.FindByNume(nume);
    }

    public void AdaugaEchipa(Echipa echipa)
    {
        _echipaRepository.Adauga(echipa);
    }

    public List<Cursa> FindAllCurse()
    {
        return _cursaRepository.FindAll();
    }

    public List<CursaDTO> ListDtoCurse()
    {
        var curse = _cursaRepository.FindAll();
        var dtoList = new List<CursaDTO>();

        foreach (var cursa in curse)
        {
            int nrPart = _participantCursaRepository.CountByCursa(cursa);
            dtoList.Add(new CursaDTO(
                cursa.Nume,
                cursa.CapacitateMinima,
                cursa.CapacitateMaxima,
                nrPart
            ));
        }

        return dtoList;
    }

    public int FindNrPartByCursa(Cursa cursa)
    {
        return _participantCursaRepository.CountByCursa(cursa);
    }

    public void AdaugaParticipantLaCursa(ParticipantCursa participantCursa)
    {
        _participantCursaRepository.adaugainbaza(participantCursa);
        NotifyParticipantAdded(participantCursa);
    }

    public List<ParticipantCursa> FindAllParticipantCursa()
    {
        return _participantCursaRepository.FindAll();
    }

    private void NotifyParticipantAdded(ParticipantCursa participantCursa)
    {
        foreach (var observerList in loggedClients.Values)
        {
            foreach (var obs in observerList)
            {
                ThreadPool.QueueUserWorkItem(_ =>
                {
                    try
                    {
                        obs.Adauga(participantCursa);
                    }
                    catch (Exception e)
                    {
                        Console.WriteLine("Error notifying observer: " + e.Message);
                    }
                });
            }
        }
    }
}

File: AN_2/SEM_2/MPP/java/lab6/Server/src/main/java/ro/mpp2024/server/ServicesImpl.java
package ro.mpp2024.server;

import ro.mpp2024.model.*;
import ro.mpp2024.persistence.*;
import ro.mpp2024.services.*;

import java.util.*;
import java.util.concurrent.*;

public class ServicesImpl implements IServices {

    private final CursaRepositoryInterface cursaRepository;
    private final EchipaRepositoryInterface echipaRepository;
    private final ParticipantRepositoryInterface participantRepository;
    private final ParticipantCursaRepositoryInterface participantCursaRepository;
    private final UtilizatorRepositoryInterface utilizatorRepository;

    private final Map<String, List<IObserver>> loggedClients = new ConcurrentHashMap<>();

    public ServicesImpl(UtilizatorRepositoryInterface utilizatorRepository, EchipaRepositoryInterface echipaRepository, ParticipantRepositoryInterface participantRepository, CursaRepositoryInterface cursaRepository, ParticipantCursaRepositoryInterface participantCursaRepository) {
        this.utilizatorRepository = utilizatorRepository;
        this.echipaRepository = echipaRepository;
        this.participantRepository = participantRepository;
        this.cursaRepository = cursaRepository;
        this.participantCursaRepository = participantCursaRepository;
    }

    @Override
    public synchronized void login(Utilizator utilizator, IObserver client) throws Exception {
        if (utilizator == null)
            throw new Exception("Utilizator invalid.");

        Utilizator found = utilizatorRepository.findByNume(utilizator.getNume());
        if (found == null || !found.verificaParola(utilizator.getParola())) {
            throw new Exception("Autentificare esuata.");
        }

        loggedClients.putIfAbsent(utilizator.getNume(), new CopyOnWriteArrayList<>());
        List<IObserver> observers = loggedClients.get(utilizator.getNume());

        if (observers.contains(client)) {
            throw new Exception("Observer deja logat.");
        }

        observers.add(client);
    }

    @Override
    public synchronized void logout(Utilizator utilizator, IObserver client) {
        if (utilizator == null) return;

        List<IObserver> observers = loggedClients.get(utilizator.getNume());
        if (observers != null) {
            observers.remove(client);
            if (observers.isEmpty()) {
                loggedClients.remove(utilizator.getNume());
            }
        }

        System.out.println("Utilizator delogat: " + utilizator.getNume());
    }

    @Override
    public synchronized Participant findByNume(String nume) {
        return participantRepository.findByNume(nume);
    }

    @Override
    public synchronized void inscriereParticipant(Participant participant) throws Exception {
        participantRepository.adauga(participant);

        List<Cursa> curse = cursaRepository.findAll();
        for (Cursa cursa : curse) {
            if (cursa.getCapacitateMinima() <= participant.getCapacitateMotor() &&
                    participant.getCapacitateMotor() <= cursa.getCapacitateMaxima()) {

                ParticipantCursa participantCursa = new ParticipantCursa(participant, cursa);
                participantCursaRepository.adaugainbaza(participantCursa);

                notifyParticipantAdded(participantCursa);
                break;
            }
        }
    }

    @Override
    public synchronized List<ParticipantDTO> findAllParticipanti() throws Exception {
        return participantRepository.findAllDTO();
    }

    @Override
    public synchronized List<Participant> cautaParticipantiDupaEchipa(String numeEchipa) throws Exception {
        Echipa echipa = echipaRepository.findByNume(numeEchipa);
        if (echipa == null) throw new Exception("Echipa nu exista");
        return participantRepository.findByEchipa(echipa);
    }

    @Override
    public synchronized Echipa findEchipaByNume(String nume) throws Exception {
        return echipaRepository.findByNume(nume);
    }

    @Override
    public synchronized void adaugaEchipa(Echipa echipa) throws Exception {
        echipaRepository.adauga(echipa);
    }

    @Override
    public synchronized List<Cursa> findAllCurse() throws Exception {
        return cursaRepository.findAll();
    }

    @Override
    public synchronized List<CursaDTO> listDtoCurse() throws Exception {
        List<Cursa> curse = cursaRepository.findAll();
        List<CursaDTO> dtoList = new ArrayList<>();
        for (Cursa cursa : curse) {
            int nrPart = participantCursaRepository.findNrPartByCursa(cursa);
            dtoList.add(new CursaDTO(cursa.getNume(), cursa.getCapacitateMinima(), cursa.getCapacitateMaxima(), nrPart));
        }
        return dtoList;
    }

    @Override
    public synchronized int findNrPartByCursa(Cursa cursa) throws Exception {
        return participantCursaRepository.findNrPartByCursa(cursa);
    }

    @Override
    public synchronized void adaugaParticipantLaCursa(ParticipantCursa participantCursa) throws Exception {
        participantCursaRepository.adaugainbaza(participantCursa);
        notifyParticipantAdded(participantCursa);
    }

    @Override
    public synchronized List<ParticipantCursa> findAllParticipantCursa() throws Exception {
        return participantCursaRepository.findAll();
    }

    private void notifyParticipantAdded(ParticipantCursa participantCursa) {
        for (List<IObserver> observerList : loggedClients.values()) {
            for (IObserver obs : observerList) {
                new Thread(() -> {
                    try {
                        obs.adauga(participantCursa);
                    } catch (Exception e) {
                        System.out.println("Error notifying observer: " + e.getMessage());
                    }
                }).start();
            }
        }
    }
}
